src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs

```diff
@@ -384,51 +385,35 @@ private Saml2SecurityToken ValidateSignature(Saml2SecurityToken samlToken, strin
 
             bool keyMatched = false;
             IEnumerable<SecurityKey> keys = null;
-            if (validationParameters.IssuerSigningKeyResolver != null)
-                keys = validationParameters.IssuerSigningKeyResolver(token, samlToken, samlToken.Assertion.Signature.KeyInfo?.Id, validationParameters);
-            else
-            {
-                var key = ResolveIssuerSigningKey(token, samlToken, validationParameters);
-                if (key != null)
-                {
-                    // remember that key was matched for throwing exception SecurityTokenSignatureKeyNotFoundException
-                    keyMatched = true;
-                    keys = new List<SecurityKey> { key };
-                }
-            }
-
-            if (keys == null)
-            {
-                // control gets here if:
-                // 1. User specified delegate: IssuerSigningKeyResolver returned null
-                // 2. ResolveIssuerSigningKey returned null
-                // Try all the keys. This is the degenerate case, not concerned about perf.
-                keys = GetAllSigningKeys(validationParameters);
-            }
+            keys = ResolveIssuerSigningKey(token, samlToken, validationParameters, out keyMatched);
 
             // keep track of exceptions thrown, keys that were tried
             var exceptionStrings = new StringBuilder();
             var keysAttempted = new StringBuilder();
             bool canMatchKey = samlToken.Assertion.Signature.KeyInfo != null;
-            foreach (var key in keys)
+
+            if (keys != null)
             {
-                try
+                foreach (var key in keys)
```

Comment:
I think this loop would be better to put in Signature.Verify() and pass the key collection instead of one key at a time. The logic to try multiple keys is needed when other signed XML blocks are verified too (complete SAML2P responses, metadata). If that logic is in the security token handler, it will have to be recreated in other places.